prereise/gather/solardata/nsrdb/nrel_api.py

```diff
@@ -142,17 +159,37 @@ def download(url):
                 raise Exception(f"Request failed: status_code={resp.status_code}")
             return resp
 
-        resp = download(url)
+        def format_to_psm3data(resp):
+            info = pd.read_csv(BytesIO(resp.content), nrows=1)
+            data_resource = pd.read_csv(BytesIO(resp.content), dtype=float, skiprows=2)
 
-        info = pd.read_csv(BytesIO(resp.content), nrows=1)
-        data_resource = pd.read_csv(BytesIO(resp.content), dtype=float, skiprows=2)
+            tz, elevation = info["Local Time Zone"], info["Elevation"]
 
-        tz, elevation = info["Local Time Zone"], info["Elevation"]
-
-        if dates is not None:
-            data_resource.set_index(
-                dates + timedelta(hours=int(tz.values[0])), inplace=True
+            if dates is not None:
+                data_resource.set_index(
+                    dates + timedelta(hours=int(tz.values[0])), inplace=True
+                )
+            return Psm3Data(
+                float(lat), float(lon), float(tz), float(elevation), data_resource
             )
-        return Psm3Data(
-            float(lat), float(lon), float(tz), float(elevation), data_resource
-        )
+
+        Psm3Data.check_attrs(attributes)
+        if cache_dir is not None:
```

Comment:
I think this would be equivalent but with a few fewer LOC:
```
    # same as before
    except FileNotFoundError:
        pass
url = self._build_url(lat, lon, attributes, year, leap_day)
resp = download(url)
psm3_data = format_to_psm3data(resp)
if cache_dir is not None:
    with open(filepath, "wb") as f:
        pickle.dump(psm3_data, f)
return psm3_data

```

Reply:
Nice! Done.